Initial webpack configuration

ggetz opened the pull request:
First revision of suggested webpack configuration for a cesium
application.

Includes a `README` with instructions for running the app, as well as other
optional configurations.

Additionally includes a travis config file that simply runs the build
step, and fails if there are compilation errors.

ggetz committed:
Initial webpack configuration

First revision of suggested webpack configuration for a cesium
application.

Includes a README with instructions for running the app, as well as other
optional configurations.

Additionally includes a travis config file that simply runs the build
step, and fails if there are compilation errors.

ggetz commented on webpack.config.js:
```diff
@@ -0,0 +1,63 @@
+const CopyWebpackPlugin = require('copy-webpack-plugin');
+
+// The path to the ceisum source code
+const cesiumSource = 'node_modules/cesium/Source';
```
To make this compile currently, I point this path to my cesium repo `../cesium/Source`, and checkout this branch with the relevant cesium changes: https://github.com/AnalyticalGraphicsInc/cesium/pull/5861

ggetz commented:
@mramato @markerikson @bampakoa Seem to be interested parties, please provide input if you have it!

In addition to what I currently have, I'd like to 1) separate cesium out into it's own bundle (maybe with the DLL plugin?), and 2) remove the (now irrelevant) warning that webpack throws about require.toUrl even though it'll never be called. However, this are changes that don't affect how the cesium library itself is compiled, and I would probably include those in the README as optional configurations.

Thanks!

bampakoa commented on README.md:
```diff
@@ -1,2 +1,109 @@
+##### ES6 Style Import
+
+ 	import Cesium from 'cesium/Cesium';
+ 	var viewer = new Cesium.Viewer('cesiumContainer');
```
Duplicate entry with line 15 above. We could leave one of them.

bampakoa commented on webpack.config.js:
```diff
@@ -0,0 +1,63 @@
+			test: /\.css$/,
+			use: [ 'style-loader', 'css-loader' ]
+		}, {
+			test: /\.(png|gif|jpg|jpeg|svg|xml|json)$/,
```
We could use [url-loader](https://github.com/webpack-contrib/url-loader) for images

ggetz replied:
True. It looks like the file-loader re-emits as well as returns the filename, which we don't need to do since we're copying over the assets.

bampakoa replied:
Maybe leave file loader for svg, xml and json?

ggetz replied:
I was able to load in json and svg files just fine with the url-loader. If it's a concern, we can make the test only include files in the `Widgets` or `Assets` directories, as that's where the one's we're serving are coming from.

bampakoa replied:
Is it possible that json, xml and svg files are large enough so that they can benefit from the DataURL conversion of url-loader? If not, then I would not mind to test them with file-loader.

bampakoa commented on webpack.config.js:
```diff
@@ -0,0 +1,63 @@
+    ],
+	node: {
+		// Resolve node module use of fs
+		fs: "empty"
```
```
    "global": true,
    "crypto": "empty",
    "tls": "empty",
    "net": "empty",
    "process": true,
    "module": false,
    "clearImmediate": false,
    "setImmediate": false
```

ggetz replied:
`fs: "empty"` was to resolve one problematic third-party error. So to keep things minimal, I don't think we need to include this whole list unless there is a definite reason.

bampakoa commented:
@ggetz We could export Cesium in its own bundle using [CommonChunksPlugin](https://webpack.js.org/plugins/commons-chunk-plugin/). Also, should we add plugin for source maps?

ggetz commented:
Thanks for looking at this @bampakoa ! 

I was looking at the [dll-plugin](https://webpack.js.org/plugins/dll-plugin/) to split out cesium from the app bundle instead of the CommonChunksPlugin.

I  didn't want to include the source map plugin by default, as ideally this is the minimal setup that developers can add to. I think we'll include a reference in the README with a suggestion to use one.

ggetz committed:
file-loader to url-loader

ggetz committed:
README updates

bampakoa commented on README.md:
```diff
@@ -38,15 +37,23 @@ However, you may want to use a different version of cesium, like if you've clone
-You could even point the cesium to a branch in github or another url in your `package.json` file.
+You could even point the cesium to a branch in GitHub or another url in your `package.json` file.
+
+### Source Maps
```
Should we also mention SourceMapDevToolPlugin? What do you think?

ggetz replied:
The documentation I linked to above links to that plugin. Unless it requires some specific configuration in order to work with Cesium, I'd say we should leave it out.

bampakoa replied:
Ooops! My bad! You are right, I did not notice it :smirk:

ggetz committed:
Fix syntax error

ggetz committed:
CommonChunksPlugin

ggetz commented:
@bampakoa I did go ahead and use the CommonChunksPlugin because the performance was a lot better with just a minimal configuration, as opposed to the dllPlugin which does give you more control, but the configuration is more involved.

bampakoa commented on webpack.config.js:
```diff
@@ -0,0 +1,63 @@
+            use: [ 'file-loader' ]
+		}]
+	},
+	plugins: [
```
new NoEmitOnErrorsPlugin() ?

bampakoa commented:
What do you think of adding NoEmitOnErrorsPlugin() ?

ggetz commented:
@bampakoa I think that may be something best left up to the individual developer to add. Maybe we can also maintain a list of possible additional plugins at the bottom of the README? That way we can keep the minimal setup and still list some of the plugins Cesium developers find useful.

bampakoa commented:
@ggetz I think NoEmitOnErrorsPlugin() should be added by default since it will prevent Cesium from being emitted if there are any errors to itself, if I am getting it right.

ggetz commented:
@bampakoa I think it will only not emit if there are compilation errors, which we are trying to avoid here entirely 😆 

Also I think it is up to the developer whether or not they'd like the bundle to still emit. I know my preference is to use the developer tools in the browser to inspect what might have gone wrong.

hpinkos commented on package.json:
```diff
@@ -0,0 +1,30 @@
+    "html-webpack-plugin": "^2.30.1",
+    "style-loader": "^0.18.2",
+    "url-loader": "^0.6.2",
+    "webpack": "^3.5.6"
```
You're missing `webpack-dev-server`

ggetz committed:
package.json fixes, fix dev server options

mramato commented:
You'll probably want to set a `.gitattributes` file as well.  Use the one from Cesium master.

mramato commented:
Should we have `npm run build` build the minified version? Or is that outside the scope of webpack?  It would be nice to show users how to build the an actual dist release with minified and pragmas removed.

mramato commented on README.md:
```diff
@@ -1,2 +1,116 @@
+
+##### Ignore
+
+Since Cesium is such a large library, if you are requiring the entire Cesium object and not individual modules, it's recommended to ignore unused parts of the Cesium library by using the [`IgnorePlugin`](https://webpack.js.org/plugins/ignore-plugin/) included with webpack.
```
We should avoid saying things like `Cesium is such a large library`; it has negative connotation.  Cesium is only as large as it needs to be to accomplish its tasks.  Also, why wouldn't you use Cesium as individual modules?  If you use it as a single module, will treeshaking remove the unused parts?  We need a better story here or we should just remove this whole Ignore section.

ggetz replied:
It seems like people are "used to" using Cesium as the global object, so if you set it up that way, it will include all the modules. I can recommend against that though. If you want to be able to copy/paste Sandcastle and other example code, using the Cesium object is the most straightforward.

You need to use ES2015 module syntax in order for tree shaking to work unfortunately.

ggetz replied:
https://webpack.js.org/guides/tree-shaking/

mramato commented on README.md:
```diff
@@ -1,2 +1,116 @@
+
+##### Removing pragmas
+
+To remove pragmas like a traditional cesium release build, use the [`webpack-strip-block`](https://www.npmjs.com/package/webpack-strip-block) plugin.
```
This should be part of the default "release" configuration (and uglify below), it's a critical best practice for deployments and many people will just use what we provide out of the box.

mramato commented on README.md:
```diff
@@ -1,2 +1,116 @@
+
+Since Cesium is such a large library, if you are requiring the entire Cesium object and not individual modules, it's recommended to ignore unused parts of the Cesium library by using the [`IgnorePlugin`](https://webpack.js.org/plugins/ignore-plugin/) included with webpack.
+
+For example, if you are not using default Assets, prevent them from being included in the bundle:
```
Perhaps have a more specific example.  Blanket ignoring assets is not a great idea.  Perhaps the Natural Earth imagery is a better option.

ggetz commented:
> Should we have npm run build build the minified version? Or is that outside the scope of webpack? It would be nice to show users how to build the an actual dist release with minified and pragmas removed.

@mramato It's easy to create another webpack config file, like `release.webpack.config.js` with the production recommendations and add anther `release` script to build with that. I don't think that's out of scope.

ggetz committed:
Update README, add release config and script

mramato commented on package.json:
```diff
@@ -0,0 +1,32 @@
+  "homepage": "https://cesiumjs.org",
+  "license": "ISC",
+  "devDependencies": {
+    "cesium": "^1.37.0",
```
Up to you, but since I merged your Cesium PR, you can now just use `http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/master/cesium-1.38.0.tgz` here and everything will work out of the box.  We can then update it to `1.38` as soon as the release is out.

ggetz replied:
Sounds good.

ggetz committed:
Cleanup, resolve cesium and loader in package.json, release config

ggetz committed:
Add meta content to README

ggetz committed:
Add image

ggetz commented:
@mramato Updated, this should be ready unless you have any more comments.

ggetz commented:
@mramato Could you take a look at this today if possible? We'd like to get it merged if it's done.

ggetz committed:
Added the numbers

mramato commented:
Is there an easy way to run the dist version  of the app?  At the very least it's good for testing.

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+
+##### ES6 Style Import
+
+ 	import Cesium from 'cesium/Cesium';
```
Is there a way to require a specific module ES6 style?

ggetz replied:
Yes, I'll add it

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+|     | `build` | `release` |
+| --- | --- | --- |
+| `Cesium` object | 9.91 MB | 2.82 MB |
+| modules | 7.43 MB | 1.91 MB |
```
Also provide gzipped values for these numbers (and label them as such) that's the value that really mattes in the end.

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+
+## Bundle size comparisons
+
+Here is a comparison of the Size of the separated `cesium.js` bundle for `release` and `build` configurations, for using the global `Cesium` object and including individual modules for the Hello World viewer.
```
These numbers are way to important to bury at the bottom here.  You should move them into the `Requiring Cesium in your application` section immediately after the require examples and explain that we officially recommend you require as modules and explain why.  Perhaps add a comment to the example code as well explaining that we are requiring in the entire module for simplicity, but production applications should consider requiring in individual modules.

mramato replied:
Actually, now that I'm reading this again, I would move these numbers to the top of the performance section, instead of linking here from there

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+* `npm run release` - Runs an optimized webpack build with `release.webpack.congif.js`
+
+### Requiring Cesium in your application
+ 
```
We should provide an intro paragraph here with some context.  Having a bunch of different options with no explantation may confuse a lot of inexperienced developers.

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+
+The following optimizations are recommended for building for production and will increase performance and result in smaller bundle sizes.
+
+For best performance, make sure you are requiring individual modules from Cesium instead of the global Cesium object. Additionally, only copy the static assets that your app requires with the `CopyWebpackPlugin` by taking advantage of the [pattern options](https://github.com/webpack-contrib/copy-webpack-plugin#pattern-properties).
```
> For best performance, make sure you are requiring individual modules from Cesium instead of the global Cesium object. 

This should be in the modules section itself.

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+
+### Performance Configurations 
+
+The following optimizations are recommended for building for production and will increase performance and result in smaller bundle sizes.
```
It's not clear to me from reading this if the below optimizations are included in the example or not (I assume they are), but the wording doesn't make that concrete enough.

mramato commented:
@ggetz I think that's all I have.  Thanks again.

Perhaps someone from @AnalyticalGraphicsInc/outreach should also read through the README once you address my last set of comments.

mramato commented on README.md:
```diff
@@ -1,2 +1,157 @@
+### Running this application
+
+	npm install
+	npm start
```
For some reason on my machine `npm start` opens a browser but doesn't navigate to the app.  If this is a webpack feature, maybe there's a bug?  If we're doing it ourselves, we should probably fix it.

ggetz replied:
I'm just passing the `open` flag to the webpack dev server. I can just remove it, it sounds like it would be easier than fixing /working around a bug.

mramato commented on .travis.yml:
```diff
@@ -0,0 +1,6 @@
+language: node_js
+node_js:
+  - '8'
```
Should we build on 6 as well (since you can do both with travis).  What is the minim node version this repo supports?

ggetz replied:
_most_ versions of node 5 should work, but it should definitely work with everything 6 and up. I'll include 6, should I also include 7?

mramato replied:
5 and 7 are both short-lived dev versions.  We only really need to worry about the LTS releases (6 and soon to be 8) (4 is technically LTS, but almost over).

ggetz commented:
> Is there an easy way to run the dist version of the app? At the very least it's good for testing.

You'd just need to serve statically serve those files somehow.

ggetz committed:
Update README, remove open flag, add node versions to .travis.yml

bampakoa commented on README.md:
```diff
@@ -9,13 +9,23 @@ The minimal recommended setup for an application using [Cesium](https://cesiumjs
 
+##### Configurations
+
+We've included two webpack configuration files in this repository. `webpack.config.js` contains the minimal recommended configuration for getting setup, and configuration for running the development server. `release.webpack.config.js` contains  an optimized configuration for production use.
```
@ggetz I would prefer `webpack.release.config.js` or `webpack.prod.config.js` for easier discoverability from the user and from tooling :smirk:

ggetz committed:
Add gzipped numbers, rename to webpack.release.config.js

ggetz commented:
@bampakoa Makes sense, renamed to `webpack.release.config.js`.

Thanks for the review @mramato, updated! @AnalyticalGraphicsInc/outreach could you take a look at the `README.md`?

ggetz committed:
Remove node v7 build

mramato commented:
> You'd just need to serve statically serve those files somehow.

Yeah, what I meant was there's no built in way to do this like we have with Cesium?  That's okay for now, we can always change it later if it becomes a common question.

mramato commented on package.json:
```diff
@@ -0,0 +1,34 @@
+    "url": "https://www.agi.com"
+  },
+  "homepage": "https://cesiumjs.org",
+  "license": "ISC",
```
This should be `Apache-2.0`

ggetz committed:
Updated licence, added script for serving release

ggetz committed:
Added info about `serve-release`

ggetz committed:
Add contentBase

ggetz commented:
> Yeah, what I meant was there's no built in way to do this like we have with Cesium? That's okay for now, we can always change it later if it becomes a common question.

We can run it with the development server like we do for the dev build. I added a `serve-release` target. It requires an additional flag to avoid adding the dev server options to the release config. 

@mramato

mramato commented:
@AnalyticalGraphicsInc/outreach are you going to be able to take a look at this soonish?  Can you please confirm either way because if not we'll merge this as-is for now.  Thanks.

vanessamoore commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+### Source Maps
+
+Enable source maps in development for easier debugging. They are only available when using the development server. There are many [options](https://webpack.js.org/configuration/devtool/). Suggested would be `eval` for fast build and rebuilt time and allowing evaluation of the webpack generated code.
```
@mramato Should this read "Suggested would be 'eval' for fast build and rebuild time..."? "rebuild" vs "rebuilt"

vanessamoore commented:
@mramato I've reviewed the file for typos etc. but might not be the best person to review in terms of content. I've talked to Sarah and she should be able to review later today. Sorry for the delay!

ggetz committed:
Fix typo in README

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
 # cesium-webpack-example
-The minimal recommended setup for an application using Cesium with Webpack.
+
+The minimal recommended setup for an application using [Cesium](https://cesiumjs.org/) with [Webpack](https://webpack.js.org/concepts/).
```
minimal => minimum

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+##### Configurations
+
+We've included two webpack configuration files in this repository. `webpack.config.js` contains the minimal recommended configuration for getting setup, and configuration for running the development server. `webpack.release.config.js` contains  an optimized configuration for production use.
```
minimal => minimum
setup, and configuration => set up and the configuration

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+Using either the `build` or `release` configurations provided, there are several ways to include Cesium in your application. There is the [CommonJS](http://requirejs.org/docs/commonjs.html) style syntax which uses `require`, and the newer [ES6](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import) style syntax which uses the `import` keyword. Both are supported by w ...
+
+You can also include all modules under the global Cesium object using `Cesium.js', or the individual modules. Requiring individual modules is preferred, resulting in smaller bundle sizes and better performance in your application. See the [Performance Configurations](#performance-configurations) section for more information on improving performance in your application.
```
`Cesium.js' => `Cesium.js`

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+	var Color = require('cesium/Core/Color');
+	var c = Color.fromRandom();
+
+or with ES6 Syntax:
```
or with ES6 Syntax:
=>
##### Import specific modules from Cesium with ES6 syntax

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+ 	import Cesium from 'cesium/Cesium';
+ 
+##### Require Cesium static asset files
```
Should we provide the ES6 syntax for this section, too?

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+ 	require('cesium/Assets/Textures/pin.svg');
+
+##### Require specific modules from Cesium
```
> ##### Require specific modules from Cesium

=>

##### Require specific modules from Cesium with CommonJS

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+### Using another Cesium location
+
+We've set the cesium Source location to be what's included with the cesium npm module.
```
"cesium Source" should be either `cesiumSource` or Cesium source

cesium npm => Cesium npm

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+	var cesiumSource = path.resolve(__dirname, 'node_modules/cesium/Source');
+
+However, you may want to use a different version of cesium, like if you've cloned the cesium source code directly. Just set the cesium location to the appropriate path.
```
> However, you may want to use a different version of cesium, like if you've cloned the cesium source code directly. Just set the cesium location to the appropriate path.

=>

If, however, you want to use a different version of Cesium&mdash;for example, if you've cloned the Cesium source code directly&mdash;just set the Cesium location to the appropriate path.

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+	var cesiumSource = path.resolve(__dirname, '../path/to/cesium/Source');
+
+You could even point the cesium to a branch in GitHub or another url in your `package.json` file.
```
> point the cesium to a branch

=>

point to a branch

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+### Performance Configurations 
+
+The following optimizations are recommended for building for production and will increase performance and result in smaller bundle sizes. An example of these configuration can be found in `webpack.release.config.js`.
```
configuration => configurations

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+The following optimizations are recommended for building for production and will increase performance and result in smaller bundle sizes. An example of these configuration can be found in `webpack.release.config.js`.
+
+For best performance, make sure you are requiring individual modules from Cesium instead of the global Cesium object. Additionally, only copy the static assets that your app requires with the `CopyWebpackPlugin` by taking advantage of the [pattern options](https://github.com/webpack-contrib/copy-webpack-plugin#pattern-properties).
```
only copy => copy only

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+##### Removing pragmas
+
+To remove pragmas like a traditional cesium release build, use the [`strip-pragma-loader`](https://www.npmjs.com/package/strip-pragma-loader).
```
like a traditional cesium => such as a traditional Cesium

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+Navigate to `localhost:8080`.
+
+##### Available Scripts
```
Scripts => scripts

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+You can also include all modules under the global Cesium object using `Cesium.js', or the individual modules. Requiring individual modules is preferred, resulting in smaller bundle sizes and better performance in your application. See the [Performance Configurations](#performance-configurations) section for more information on improving performance in your application.
+ 
+##### CommonJS Require
```
Require => require

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+ 	var Cesium = require('cesium/Cesium');
+ 	var viewer = new Cesium.Viewer('cesiumContainer');
+
+##### ES6 Style Import
```
Style Import => style import

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+You could even point the cesium to a branch in GitHub or another url in your `package.json` file.
+
+### Source Maps
```
Maps => maps

slchow commented on README.md:
```diff
@@ -1,2 +1,171 @@
+
+	devTool: `eval`
+
+### Performance Configurations 
```
Configurations => configurations

slchow commented on README.md:
````diff
@@ -1,2 +1,171 @@
+}]
+```
+
+##### Uglify and Minify
````
Minify => minify

slchow reviewed (commented):
Looks great! All suggested changes are minor.

ggetz committed:
copyedits

ggetz commented:
Thanks @slchow , much appreciated!

mramato commented:
Thanks everyone!  Awesome job @ggetz!
